### ISSUE TITLE
Checkpointing Inefficiency

### ISSUE THREAD

**Issue by AntonioCarta** [ContinualAI/avalanche#1261]
I'm using the checkpointing for the first time. I'm opening this issue to keep track of the memory/time inefficiencies that I found on a nontrivial usecase.

Usecase: Split TinyImageNet, VGGWide, LwF-like strategy

note: time and space are measured using dumps (strings). The binary method dump is a bit more efficient but not much more (total memory occupation is half of dumps):
```
Mapping cuda:0 to cuda:0
strategy 	 634 MB  - serialization time:  99.80834984779358
rng_manager 	 0 MB  - serialization time:  0.014878034591674805
exp_counter 	 0 MB  - serialization time:  6.651878356933594e-05
model 	 35 MB  - serialization time:  0.1376795768737793
device 	 0 MB  - serialization time:  0.00024056434631347656
plugins 	 37 MB  - serialization time:  0.22279644012451172
experience 	 6 MB  - serialization time:  32.81073069572449
is_training 	 0 MB  - serialization time:  0.00026869773864746094
current_eval_stream 	 6 MB  - serialization time:  32.77020788192749
optimizer 	 104 MB  - serialization time:  0.20015549659729004
_criterion 	 0 MB  - serialization time:  0.0024137496948242188
train_epochs 	 0 MB  - serialization time:  7.462501525878906e-05
train_mb_size 	 0 MB  - serialization time:  6.842613220214844e-05
eval_mb_size 	 0 MB  - serialization time:  6.723403930664062e-05
evaluator 	 0 MB  - serialization time:  0.028775691986083984
eval_every 	 0 MB  - serialization time:  7.414817810058594e-05
clock 	 0 MB  - serialization time:  0.00067901611328125
adapted_dataset 	 0 MB  - serialization time:  2.9171347618103027
dataloader 	 0 MB  - serialization time:  3.1857316493988037
mbatch 	 4 MB  - serialization time:  0.004255771636962891
mb_output 	 0 MB  - serialization time:  0.0011248588562011719
loss 	 0 MB  - serialization time:  0.0009379386901855469
_stop_training 	 0 MB  - serialization time:  7.271766662597656e-05
temperature 	 0 MB  - serialization time:  7.033348083496094e-05
ood_dataset 	 0 MB  - serialization time:  0.8863067626953125
init_from_first 	 0 MB  - serialization time:  0.00012493133544921875
classifier_name 	 0 MB  - serialization time:  7.987022399902344e-05
basedir 	 0 MB  - serialization time:  7.104873657226562e-05
expert_epochs 	 0 MB  - serialization time:  6.175041198730469e-05
expert_strategy 	 0 MB  - serialization time:  0.07527446746826172
_cm 	 0 MB  - serialization time:  0.0003135204315185547
prev_model 	 35 MB  - serialization time:  0.10101175308227539
_eval_streams 	 414 MB  - serialization time:  91.48722577095032
expert_model 	 34 MB  - serialization time:  0.12400960922241211
LatentKDProjection 	 1 MB  - serialization time:  0.12815618515014648
CustomEarlyStoppingPlugin 	 35 MB  - serialization time:  0.07553935050964355
CheckpointPlugin 	 0 MB  - serialization time:  0.004965066909790039
import sys; print('Python %s on %s' % (sys.version, sys.platform))
EvaluationPlugin 	 0 MB  - serialization time:  0.02569103240966797
PeriodicEval 	 0 MB  - serialization time:  0.0005555152893066406
Clock 	 0 MB  - serialization time:  0.0005908012390136719
```

It seems that the serialization has issues with highly recursive objects with circular dependencies like `_eval_streams`. The good news is that we don't need to checkpoint those objects, the bad news is that right now we are saving them. Other expensive objects like `optimizer` can often be dropped this depend on the strategy.

**Comment by AntonioCarta** [comment]
I'm thinking about an alternative API for checkpointing. Right now we do:
**Save**
```
save(strategy)
```
**Load**
```
strat = load(strat)
if strat is not None:
    strat = Strat(...)
```

What we need is a method to update an object with only the dynamic state (the attribute that we really need to save):
**Save**:
```
save(strat)
```
**Load**
```
strat = Strat(...)
strat, eid = maybe_load(strat, fname)
```
With this method, static attributes are initialized by the constructor, while dynamic ones are overridden by the checkpoint.
This is similar to pytorch `save_state_dict`, `load_state_dict`. I don't think pickle/dill allow this kind loading mechanism.

I have a very early draft of this that I'm working on because I need it for my experiments but I can push it here when it's ready.

**Comment by lrzpellegrini** [comment]
I agree some elements may be skipped. It's mostly a matter of identifying these elements and then skipping them by implementing `__set/getstate__` where needed. About the optimizer, that is very strategy specific...

One possible solution for those recurring elements for which we do not know if they can be dropped (like the optimizer) is to allow the user to explicitly tell the plugin to drop them.

Regarding the benchmark-related objects (experiences, streams, ...), I think it's ok to drop them by default. My idea is that, for these objects, we can store their identifier only (originating stream + ID + ...) and then, at unpickling time, replace them to the benchmark object instantiated by the user at the beginning of the stream.

However, I can also foresee users storing a checkpoint on a benchmark and then re-load the last checkpoint to confront a completely different benchmark (but by keeping replay elements and other old-benchmark-related elements). Maybe we should also allow for this.

**Comment by AntonioCarta** [comment]
Fixed by #1338
